File: python/자료구조/더 맵게.py

""" Level 2. 더 맵게 """

from heapq import heappush, heappop, heapify

def solution(scoville, K):
    
    answer = 0
    slen = len(scoville)

    heapify(scoville)

    while scoville[0] < K and slen > 1:

        num1 = heappop(scoville)
        num2 = heappop(scoville)
        heappush(scoville, num1 + num2 * 2)

        answer += 1
        slen -= 1

    return answer if scoville[0] >= K else -1

print(solution([1, 2, 3, 9, 10, 12], 7))        # 2
print(solution([1, 1, 1, 1, 1, 1], 1500))          # -1

File: Java/src/자료구조/더맵게.java
package 자료구조;

import java.util.*;

public class 더맵게 {

    public static void main(String[] args) {

        int[] s1 = {1, 2, 3, 9, 10, 12};
        int k1 = 7;
        System.out.println(solution(s1, k1));

        int[] s2 = {1, 2, 3, 9, 10, 12};
        int k2 = 5000;
        System.out.println(solution(s2, k2));
    }

    public static int solution(int[] scoville, int K) {
        
        PriorityQueue<Integer> q = new PriorityQueue<>();

        for (int sc : scoville) {
            q.add(sc);
        }

        int length = scoville.length;
        int answer = 0;
        while (length > 1) {

            if (K <= q.peek()) return answer;

            q.add(q.poll() + q.poll() * 2);
            answer += 1;
            length -= 1;
        }
        
        if (K <= q.peek()) {
            return answer;
        } else {
            return -1;
        }
    }
}
